### PR TITLE
Various Xserver/os backports

### PR DESCRIPTION
**Pull request by sunweaver**
This PR requires #331 to be merged first.

### CODE REVIEW

**Comment by sunweaver** [comment]
@Ionic: here is the bulk backport pull request... Once this one is in, we can look at the libXfont2 PR.

**Review by uli42 — COMMENTED** [review]
Hmm, this seems to affect ssh-X11 forwarding and might therefore have impact on the x2go setup (which, IIRC, uses ssh -X to trigger ssh optimizations)

**Comment by Ionic** [comment]
Github does something weird with this PR. Allegedly, 7ad1f8cc17cbdd88f53bb767aa20b5b816d1aa3a is supposed to come before 495889247c6e07b21dc2f90fd38856c7d68031ab, but that's impossible because the later introduced the change that is then changed again by the former...

**Comment by Ionic** [comment]
You somehow managed to break Github with this PR. Most commits seem to be shown out of order - but the git history looks fine. I'll create a github bug report for this...

**Comment by Ionic** [comment]
Okay, seems to be a known issue and they won't do anything about it, c.f. https://help.github.com/articles/why-are-my-commits-in-the-wrong-order/

File: nx-X11/programs/Xserver/os/connection.c [hunk 1]
```diff
@@ -988,7 +988,7 @@ CloseDownConnection(ClientPtr client)
     OsCommPtr oc = (OsCommPtr)client->osPrivate;
 
     if (FlushCallback)
-	CallCallbacks(&FlushCallback, NULL);
+	CallCallbacks(&FlushCallback, client);
```
Comment (Ionic) [inline nx-X11/programs/Xserver/os/connection.c]: This commit doesn't fit in, really... Semantically, when looking at the previous one, it does, but the commit message references a completely different upstream commit that has nothing to do with your changes.

Your changes match this upstream commit:

```
commit b380f3ac51f40ffefcde7d3db5c4c149f274246d
Author: Michel Dänzer <michel.daenzer@amd.com>
Date:   Tue Aug 2 17:53:01 2016 +0900

    dix: Pass ClientPtr to FlushCallback

    This change has two effects:

    1. Only calls FlushCallbacks when we're actually flushing data to a
       client. The unnecessary FlushCallback calls could cause significant
       performance degradation with compositing, which is significantly
       reduced even without any driver changes.

    2. By passing the ClientPtr to FlushCallbacks, drivers can completely
       eliminate unnecessary flushing of GPU commands by keeping track of
       whether we're flushing any XDamageNotify events to the client for
       which the corresponding rendering commands haven't been flushed to
       the GPU yet.

    Reviewed-by: Adam Jackson <ajax@redha.com>
    Signed-off-by: Michel Dänzer <michel.daenzer@amd.com>
```

Comment (Ionic) [reply nx-X11/programs/Xserver/os/connection.c]: If you really meant to backport the commit you mentioned in this commit message, you will have to do that as well, but I don't know if it makes sense at this point in time, since the mentioned upstream commit fixes a regression that appeared after upstream changed to its new `ospoll`-/`epoll`-based implementation. We do not have that yet.

File: nx-X11/programs/Xserver/os/Imakefile [hunk 2]
```diff
@@ -121,16 +124,51 @@ TMEMCMP_OBJS = timingsafe_memcmp.o
...
-           OBJS = WaitFor.o access.o connection.o io.o $(COLOR_OBJS) \
-                  osinit.o utils.o log.o auth.o mitauth.o secauth.o \
-                  $(XDMAUTHOBJS) $(RPCOBJS) xdmcp.o OtherObjects \
-                  xstrans.o $(SNPRINTF_OBJS) $(STRLCAT_OBJS) \
-		  $(REALLOCARRAY_OBJS) xprintf.o $(TMEMCMP_OBJS)
+           SRCS = WaitFor.c		\
+                  access.c		\
+                  connection.c		\
+                  io.c			\
+                  $(COLOR_SRCS)		\
```
Comment (Ionic) [inline nx-X11/programs/Xserver/os/Imakefile]: You've broken the escape character lineup with tabsize = 8 here and on several occasions below that line.

I guess you don't see it because you set your editor to a different tabsize, just mentioning it.

Also, the `pkg-config` line way down is too long even for tabsize = 8, so while you're at it, make sure the escape chars line up in that hunk?

File: nx-X11/programs/Xserver/os/Imakefile [hunk 3]
```diff
@@ -121,16 +124,51 @@ TMEMCMP_OBJS = timingsafe_memcmp.o
...
+                  mitauth.c		\
+                  secauth.c		\
+                  $(XDMAUTHSRCS)	\
+                  $(RPCSRCS)		\
+                  xdmcp.c		\
+                  OtherSources		\
+                  xstrans.c		\
+                  $(SNPRINTF_SRCS)	\
+                  $(STRLCAT_SRCS)	\
+                  xprintf.c		\
```
Comment (Ionic) [inline nx-X11/programs/Xserver/os/Imakefile]: Try to keep the order. In the original string, `xprintf.c` comes after `$(REALLOCARRAY_SRCS)`. This isn't a problem per se, but in the OBJS definition, the ordering is different again: first `$(REALLOCARRAY_OBJS)`, then `$(TMEMCMP_OBJS)`, then `xprintf.o`. Please try to keep both definitions in synch, since humans are pretty good at pattern matching and would expect entries to be at the same relative location if all the others are. That is, I don't care about the actual order, but I'd like to have a consistent one.

In both cases, too, `$(REALLOCARRAY_SRCS)` is followed by spaces, whereas you converted all the other lines to use tabs.

File: nx-X11/programs/Xserver/os/access.c [hunk 4]
```diff
@@ -496,7 +496,7 @@ DefineSelf (int fd)
 void
 DefineSelf (int fd)
 {
-#if !defined(TCPCONN) && !defined(UNIXCONN) && !defined(MNX_TCPCONN)
+#if !defined(TCPCONN) && !defined(UNIXCONN)
```
Comment (Ionic) [inline nx-X11/programs/Xserver/os/access.c:495]: Making a note: `MNX_TCPCONN` left over in `nx-X11/lib/X11/ConnDis.c`. It doesn't fit in with the `Xserver/os` changes, but we probably want to remove it outside of this PR as well.

File: nx-X11/programs/Xserver/os/access.c [hunk 5]
```diff
@@ -1314,57 +1323,64 @@ LocalClientCredAndGroups(ClientPtr client, int *pUid, int *pGid,
...
-	    return -1;
-    if (pUid != NULL)
-	    *pUid = peercred.uid;
-    if (pGid != NULL)
-	    *pGid = peercred.gid;
+    if (getsockopt(fd, SOL_SOCKET, SO_PEERCRED, &peercred, &so_len) == -1) {
+	FreeLocalClientCreds(lcc);
+	return -1;
+    }
+     lcc->euid = peercred.uid;
```
Comment (Ionic) [inline nx-X11/programs/Xserver/os/access.c]: Unnecessary whitespace error here?

File: nx-X11/programs/Xserver/os/connection.c [hunk 6]
```diff
@@ -534,8 +534,8 @@ AuthAudit (ClientPtr client, Bool letin,
     char addr[128];
     char *out = addr;
 
-    int client_uid;
     char client_uid_string[32];
```
Comment (Ionic) [inline nx-X11/programs/Xserver/os/connection.c]: Hm, not an immediate problem with your backport, but it looks like `client_uid_string` had size 64 in the original code.

That change was part of

```
commit a7b944f0d96c3e0e15e75378a04def1ac96089fb
Author: Alan Coopersmith <alan.coopersmith@sun.com>
Date:   Wed Nov 1 16:17:49 2006 -0800

    If getpeerucred() is available, include pid & zoneid in audit messages too
```

Backporting the whole commit doesn't make sense since it introduces `#ifdef HAS_GETPEERUCRED` blocks which are then removed again with the commit you backported now, but at least cherry-picking that one change might.

File: nx-X11/programs/Xserver/Xext/shm.c [hunk 7]
```diff
@@ -167,7 +166,7 @@ static Bool CheckForShmSyscall()
     int shmid = -1;
 
     /* If no SHM support in the kernel, the bad syscall will generate SIGSYS */
-    oldHandler = signal(SIGSYS, SigSysHandler);
+    oldHandler = OsSignal(SIGSYS, SigSysHandler);
```
Comment (Ionic) [inline nx-X11/programs/Xserver/Xext/shm.c:169]: What about `nx-X11/programs/Xserver/os/utils.c`? Missing changes for `SIGCHLD` there.

File: nx-X11/programs/Xserver/os/connection.c [hunk 8]
```diff
@@ -857,59 +859,44 @@ EstablishNewConnections(ClientPtr clientUnused, void * closure)
...
-	    ErrorConnMax(new_trans_conn);
-	}
-      }
-#ifndef WIN32
-    }
-#endif
     return TRUE;
 }
 
+static void
```
Comment (Ionic) [inline nx-X11/programs/Xserver/os/connection.c:899]: Something in this commit looks odd...

The upstream commit 7b02f0b87ec2fa0cc5a65307a1fd55c671cec884 has this hunk:

```diff
@@ -1417,8 +1398,7 @@ ListenOnOpenFD(int fd, int noxauth)
     ListenTransConns[ListenTransCount] = ciptr;
     ListenTransFds[ListenTransCount] = fd;

-    FD_SET(fd, &WellKnownConnections);
-    FD_SET(fd, &AllSockets);
+    SetNotifyFd(fd, QueueNewConnections, X_NOTIFY_READ, NULL);

     /* Increment the count */
     ListenTransCount++;
```

which is missing in your backport.

While the function in question doesn't exist in our source tree, I hope we're not missing out on the `SetNotifyFd()` call. Probably not, since the function is not even used on anything but the Darwin and Wayland DDX'.

**Review by Ionic — CHANGES_REQUESTED** [review]

File: nx-X11/programs/Xserver/os/connection.c (hunk 1 above)
Comment (sunweaver) [reply nx-X11/programs/Xserver/os/connection.c]: Oh dear. Sorry for this flaw. You are, of course, right on this. Amended locally already.

Comment (sunweaver) [reply nx-X11/programs/Xserver/os/connection.c]: There actually was another issue with this commit. I backported the above commit from 2016 and then X.org commit c65f610e12f9df168d5639534ed3c2bd40afffc8 from 2010. I re-did this commit and the 2010 commit and wove them all into this one with extra backport commit message.

File: nx-X11/programs/Xserver/os/io.c [hunk 9]
```diff
@@ -965,6 +965,10 @@ WriteToClient (ClientPtr who, int count, const void *__buf)
 	  CriticalOutputPending = FALSE;
 	  NewOutputPending = FALSE;
 	}
+
+	if (FlushCallback)
+	    CallCallbacks(&FlushCallback, NULL);
+
 	return FlushClient(who, oc, buf, count);
```
Comment (sunweaver) [inline nx-X11/programs/Xserver/os/io.c]: @Ionic: this part of the commit now becomes obsolete. Has been moved to FlushClient() function later in 2016.

Comment (sunweaver) [reply nx-X11/programs/Xserver/os/io.c]: This commit has been removed now and will appear as reference in  65b6a62bc726b5e4f4833ed1e27733e22fabcfa8.

File: nx-X11/programs/Xserver/os/Imakefile (hunk 2 above)
Comment (sunweaver) [reply nx-X11/programs/Xserver/os/Imakefile]: > You've broken the escape character lineup with tabsize = 8 here and on several occasions below that line.
>
> I guess you don't see it because you set your editor to a different tabsize, just mentioning it.
> 
> Also, the pkg-config line way down is too long even for tabsize = 8, so while you're at it, make sure the escape chars line up in that hunk?

@Ionic:I use tab size 8 here and the breakage only appears in the diff output (because of the extra "+" at the beginning of the diff line. In the file itself, the tabbing is ok.

File: nx-X11/programs/Xserver/os/Imakefile (hunk 3 above)
Comment (sunweaver) [reply nx-X11/programs/Xserver/os/Imakefile]: > Try to keep the order. In the original string, xprintf.c comes after $(REALLOCARRAY_SRCS). This isn't a problem per se, but in the OBJS definition, the ordering is different again: first $(REALLOCARRAY_OBJS), then $(TMEMCMP_OBJS), then xprintf.o. Please try to keep both definitions in synch, since humans are pretty good at pattern matching and would expect entries to be at the same relative location if all the others are. That is, I don't care about the actual order, but I'd like to have a consistent one.
>
> In both cases, too, $(REALLOCARRAY_SRCS) is followed by spaces, whereas you converted all the other lines to use tabs.

Nice catch! Fixed.

File: nx-X11/programs/Xserver/os/access.c (hunk 4 above)
Comment (sunweaver) [reply nx-X11/programs/Xserver/os/access.c:495]: > Making a note: MNX_TCPCONN left over in nx-X11/lib/X11/ConnDis.c. It doesn't fit in with the Xserver/os changes, but we probably want to remove it outside of this PR as well.

See new issue #399...

File: nx-X11/programs/Xserver/os/connection.c (hunk 6 above)
Comment (sunweaver) [reply nx-X11/programs/Xserver/os/connection.c]: Good catch. Fixed.

File: nx-X11/programs/Xserver/Xext/shm.c (hunk 7 above)
Comment (sunweaver) [reply nx-X11/programs/Xserver/Xext/shm.c:169]: @Ionic:

> What about nx-X11/programs/Xserver/os/utils.c? Missing changes for SIGCHLD there.

Oh dear, this commit has been completely redone. Basically by pulling in all recent and ancient changes to the smart scheduler ABI.

File: nx-X11/programs/Xserver/include/dixstruct.h [hunk 10]
```diff
@@ -160,7 +160,7 @@ extern Bool SmartScheduleIdle;
 extern Bool SmartScheduleTimerStopped;
 extern Bool SmartScheduleStartTimer(void);
 #ifdef NXAGENT_SERVER
-extern Bool SmartScheduleStopTimer(void);
+extern void SmartScheduleStopTimer(void);
```
Comment (sunweaver) [inline nx-X11/programs/Xserver/include/dixstruct.h]: This commit ("Xserver/include/dixstruct.h: Fix type mismatch for SmartScheduleStopTimer() prototype.") will get dropped with my next version of this PR branch.

File: nx-X11/programs/Xserver/os/connection.c (hunk 8 above)
Comment (sunweaver) [reply nx-X11/programs/Xserver/os/connection.c:899]: > [...]
>
> While the function in question doesn't exist in our source tree, I hope we're not missing out on the SetNotifyFd() call. Probably not, since the function is not even used on anything but the Darwin and Wayland DDX'.

The ListenOnOpenFD() function was specially added to the Xserver ABI for XWayland. It may be needed for nxagent later, if we want to include external socket handling support for nxagent. But for now, we should be fine.

**Comment by sunweaver** [comment]
@Ionic: ok. I think I worked myself through all your notes / requests.

The signal() ->  OsSignal() commit got fully reworked now and ships many more backports in a bundle.

Also, I added some command line parsing commit for the "-x" option at the end of the commit series.

Before merging this PR, I'd like to do some more performance tests, as I am not 100% sure, the Signal and SmartScheduler changes won't bring us performance problems.

**Comment by sunweaver** [comment]
@Ionic: ok, I did various tests. Performance seems to be slightly improve with this PR. However, it seems we need to keep an eye on drawing glitches, esp. when using shadow mode.

File: nx-X11/programs/Xserver/os/access.c (hunk 5 above)
Comment (Ionic) [reply nx-X11/programs/Xserver/os/access.c]: Ping

File: nx-X11/programs/Xserver/os/Imakefile (hunk 3 above)
Comment (Ionic) [reply nx-X11/programs/Xserver/os/Imakefile]: Okay, almost done. Line 203 has a space char after `-I.`.

File: nx-X11/programs/Xserver/Xext/shm.c (hunk 7 above)
Comment (Ionic) [reply nx-X11/programs/Xserver/Xext/shm.c:169]: Phew... this one got really huge... and is still missing a lot of commit references for quite some of the `itimer` changes.

Also, I noticed `SMART_SCHEDULE_DEFAULT_INTERVAL` and `SMART_SCHEDULE_MAX_SLICE` changed upstream. If we change the default values as well, this might influence performance. No idea if for the better or worse, though.

One commit that should have been mentioned:

```
commit e10ba9e4b52269b2ac75c4802dce4ca47d169657
Author: Keith Packard <keithp@keithp.com>
Date:   Wed Nov 11 22:02:01 2015 -0800

    Remove non-smart scheduler. Don't require setitimer.

    This allows the server to call GetTimeInMillis() after each request is
    processed to avoid needing setitimer. -dumbSched now turns off the
    setitimer.

    Reviewed-by: Adam Jackson <ajax@redhat.com>
    Signed-off-by: Keith Packard <keithp@keithp.com>
```

It also looks like this hunk is missing from `nx-X11/programs/Xserver/hw/nxagent/NXdispatch.c`:

```diff
@@ -415,7 +418,10 @@ Dispatch(void)
                    result = BadLength;
                else
                    result = (* client->requestVector[MAJOROP])(client);
-
+
+               if (!SmartScheduleSignalEnable)
+                   SmartScheduleTime = GetTimeInMillis();
+
                if (result != Success)
                {
                    if (client->noClientException != Success)
```

But wait, there's more!

In `nx-X11/programs/Xserver/hw/nxagent/Display.h`:

```diff
 #define nxagentDisableTimer() \
 \
-    if (SmartScheduleTimerStopped == 0) \
-    { \
-      SmartScheduleStopTimer(); \
-    } \
-\
-    SmartScheduleDisable = 1;
+    SmartScheduleStopTimer(); \
+    SmartScheduleSignalEnable = FALSE;

 /*
  * File descriptor currently used by
```

`SmartScheduleSignalEnable` is set unguarded. This will go boom if `HAVE_SETITIMER` is undefined, because in that case `SmartScheduleSignalEnable` is macro'd to `FALSE` and we're then effectively setting `FALSE = FALSE`.

File: nx-X11/programs/Xserver/os/WaitFor.c [hunk 11]
```diff
@@ -213,18 +213,10 @@ WaitForSomething(int *pClientsReady)
...
-		wt = &waittime;
-	    }
-	    else
-	    {
-		XFD_COPYSET (&ClientsWithInput, &clientsReadable);
-		break;
-	    }
+	    someReady = TRUE;
+	    waittime.tv_sec = 0;
+	     waittime.tv_usec = 0;
```
Comment (Ionic) [inline nx-X11/programs/Xserver/os/WaitFor.c]: Unnecessary whitespace error.

**Review by Ionic — CHANGES_REQUESTED** [review]
Not just because of the whitespace error, but we've got a few other things that aren't perfect yet...

File: nx-X11/programs/Xserver/os/Imakefile (hunk 3 above)
Comment (sunweaver) [reply nx-X11/programs/Xserver/os/Imakefile]: Ok, fixed now.

File: nx-X11/programs/Xserver/os/access.c (hunk 5 above)
Comment (sunweaver) [reply nx-X11/programs/Xserver/os/access.c]: > Unnecessary whitespace error here?

Sorry, missed this one. Fixed now

File: nx-X11/programs/Xserver/os/WaitFor.c (hunk 11 above)
Comment (sunweaver) [reply nx-X11/programs/Xserver/os/WaitFor.c]: > Unnecessary whitespace error. (in os/WaitFor.c)

Fixed.

File: nx-X11/programs/Xserver/Xext/shm.c (hunk 7 above)
Comment (sunweaver) [reply nx-X11/programs/Xserver/Xext/shm.c:169]: @Ionic: Display.h and NXdispatch.c have been amended.

**Comment by sunweaver** [comment]
@ionic: next review iteration.

**Review by Ionic — APPROVED** [review]
Everything has been fixed and I wasn't able to come up with anything new while again looking at the `OsSignal` change.
